pagefaultgames/pokerogue
# [Item] Add Weather and Terrain Extender Item

> [!NOTE]
> The sprite for the item was included in https://github.com/pagefaultgames/pokerogue/pull/5542.

## What are the changes the user will see?


Adds the functionality for `Mystical Rock`, an item that has the following properties:
- Tier: Ultra
- Max Stack: 2
- Effect: Adds two turns towards weather and terrain moves per stack.
- Spawn Conditions: The item will be pushed to the item reward pool with a weight of 10 if any Pokémon in the party have the following moves and abilities:
  - Moves
    -  Sunny Day
    - Rain Dance
    - Sandstorm
    - Snowscape
    - Hail
    - Chilly Reception
    - Electric Terrain
    - Psychic Terrain
    - Grassy Terrain
    - Misty Terrain
  - Abilities
    - Drought
    - Orichalcum Pulse
    - Drizzle
    - Sand Stream
    - Sand Spit
    - Snow Warning
    - Electric Surge
    - Hadron Engine
    - Psychic Surge
    - Grassy Surge
    - Seed Sower
    - Misty Surge

## Why am I making these changes?




Requested by the @pagefaultgames/balance-team, this original item was made to primarily add the field effect extension effects of several mainline items ([Heat Rock](https://bulbapedia.bulbagarden.net/wiki/Heat_Rock), [Damp Rock](https://bulbapedia.bulbagarden.net/wiki/Damp_Rock), [Smooth Rock](https://bulbapedia.bulbagarden.net/wiki/Smooth_Rock), [Icy Rock](https://bulbapedia.bulbagarden.net/wiki/Icy_Rock), and [Terrain Extender](https://bulbapedia.bulbagarden.net/wiki/Terrain_Extender)) into one item and to curb reward pool bloat while providing extra utility for weather-based and terrain-based strategies in a given run.

## What are the changes from a developer perspective?



The `FieldEffectModifier` class extending `PokemonHeldItemModifier` in `src/modifier/modifier.ts` was written to achieve the effect of the item.

In `src/modifier/modifier-type.ts`, a `MYSTICAL_ROCK` entry was made in the `modifierTypes` array along with an accompanying `WeightedModifierType` entry in `ModifierTier.Ultra` array within the `modifierPool` to essentially provide the spawning conditions explained above.

Both the `Arena` member functions dealing with setting weather and terrain, `trySetWeather` and `trySetTerrain`, had their parameters adjusted to take a `Pokemon` object, the user, instead of a boolean dictating whether a Pokémon had set that weather or terrain specifically. All calls to this function were also adjusted based on the context of the original function parameters.

Unit tests were added to test the functionality of the item.

### Screenshots/Videos



#### One Stack, Weather
<img width="650" alt="383003865-96cb8e23-1716-49fb-a8ae-b82e48cdcae1" src="https://github.com/user-attachments/assets/8e6944ea-c35e-494b-bd88-74ae5c0f1b19" />

#### Two Stacks, Terrain
<img width="1220" alt="383003776-142f4cc2-efba-46f1-a70a-5417fc56ac99" src="https://github.com/user-attachments/assets/b3a01332-3d69-4577-a345-2e7e4555a212" />

## How to test the changes?





For general testing related to item functionality, the src/overrides.ts file was modified such that the STARTING_HELD_ITEMS_OVERRIDE array included an entry for either Toxic Orb, { name: "TOXIC_ORB" } or Flame Orb, { name: "FLAME_ORB" }.

For general testing related to item functionality, the `overrides` variable in the `src/overrides.ts` file can be modified such that the `STARTING_HELD_ITEMS_OVERRIDE` has an entry for the `MYSTICAL_ROCK` item, which is shown below.
```ts
const overrides = {
  STARTING_HELD_ITEMS_OVERRIDE: [ { name: "MYSTICAL_ROCK", count: 2 } ]
} satisfies Partial<InstanceType<typeof DefaultOverrides>>;
```
The "Active Battle Effects" tab in-game is useful to see the effects of the item numerically along with changing the `count` field in the override entry to see the difference stack count makes.

## Checklist
  - https://github.com/pagefaultgames/pokerogue-locales/pull/51
